{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/alexi/Documents/ValDX\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexi/anaconda3/envs/HDXER_ENV/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "envs ['# conda environments:', '#', 'base                     /home/alexi/anaconda3', 'HDXER_ENV             *  /home/alexi/anaconda3/envs/HDXER_ENV', 'openfold_env             /home/alexi/anaconda3/envs/openfold_env', 'plumed-masterclass-2022     /home/alexi/anaconda3/envs/plumed-masterclass-2022', '                         /home/alexi/localcolabfold/colabfold-conda', '                         /home/alexi/localcolabfold/conda', '']\n",
      "name base\n",
      "name HDXER_ENV\n",
      "Environment 'HDXER_ENV' found.\n",
      "Path to 'HDXER_ENV' environment: /home/alexi/anaconda3/envs/HDXER_ENV\n",
      "PATH /home/alexi/anaconda3/envs/HDXER_ENV:/home/alexi/anaconda3/envs/HDXER_ENV/bin:/home/alexi/anaconda3/condabin:/home/alexi/.vscode/cli/servers/Stable-903b1e9d8990623e3d7da1df3d33db3e42d80eda/server/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin\n",
      "calc_hdx\n",
      "/home/alexi/Documents/HDXer/HDXer/calc_hdx.py\n",
      "Experiment name test_full0.5 already exists. Attempting to change name to test_full0.50\n",
      "Experiment name test_full0.5 already exists. Attempting to change name to test_full0.51\n",
      "Experiment name test_full0.5 already exists. Attempting to change name to test_full0.52\n",
      "Experiment name test_full0.5 already exists. Attempting to change name to test_full0.53\n",
      "Experiment name test_full0.5 already exists. Attempting to change name to test_full0.54\n",
      "Experiment name test_full0.5 already exists. Attempting to change name to test_full0.55\n"
     ]
    }
   ],
   "source": [
    "### ValDXer testing\n",
    "import os\n",
    "os.environ[\"HDXER_PATH\"] = \"/home/alexi/Documents/HDXer\"\n",
    "\n",
    "from ValDX.ValidationDX import ValDXer\n",
    "from ValDX.VDX_Settings import Settings\n",
    "import pandas as pd\n",
    "import MDAnalysis as mda\n",
    "from MDAnalysis.coordinates.XTC import XTCWriter\n",
    "\n",
    "from pdbfixer import PDBFixer\n",
    "from openmm.app import PDBFile\n",
    "\n",
    "settings = Settings(name='test_full0.5')\n",
    "settings.replicates = 3\n",
    "settings.gamma_range = (2,6)\n",
    "settings.train_frac = 0.5\n",
    "settings.RW_exponent = [0]\n",
    "settings.split_mode = 'R3'\n",
    "settings.stride = 1000\n",
    "# settings.HDXer_stride = 10000\n",
    "\n",
    "settings.RW_do_reweighting = False\n",
    "settings.RW_do_params = True\n",
    "import pickle\n",
    "\n",
    "VDX = ValDXer(settings)\n",
    "expt_name = 'Experimental'\n",
    "test_name = \"MBPwt1_test\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "    # ### add code to read in sequence from CIF file instead of copying it manually\n",
    "\n",
    "    # cif_file = \"/Users/alexi/Library/CloudStorage/OneDrive-Nexus365/Rotation_Projects/Rotation_3/Project/ValDX/raw_data/MBP/MaltoseBindingProtein/AF-P0AEX9-F1-model_v4.cif\"\n",
    "\n",
    "    # sequence_header = \"_entity_poly.pdbx_seq_one_letter_code\"\n",
    "    # sequence = \"\"\n",
    "    # seq_head_idx = 0\n",
    "    # with open(cif_file, 'r') as f:\n",
    "    #     lines = f.readlines()\n",
    "    #     for idx, line in enumerate(lines):\n",
    "    #         if sequence_header in line:\n",
    "    #             seq_head_idx = idx+1\n",
    "    #             break\n",
    "        \n",
    "    #     for idx, line in enumerate(lines[seq_head_idx:]):\n",
    "    #         if idx > 0 and line[0] == \";\":\n",
    "    #             break\n",
    "    #         sequence += line.strip()\n",
    "\n",
    "\n",
    "    # # print(sequence)\n",
    "\n",
    "\n",
    "\n",
    "    # # strip sequence of non letters\n",
    "    # sequence = ''.join([i for i in sequence if i.isalpha()])\n",
    "\n",
    "    # print(sequence)\n",
    "\n",
    "    # print(\"Sequence length: \", len(sequence))\n",
    "\n",
    "\n",
    "    # # convert sequence to FASTA format\n",
    "    # def write_fasta(sequence, header, file_name):\n",
    "    #     \"\"\"\n",
    "    #     Writes a single-letter amino acid sequence to a FASTA file.\n",
    "        \n",
    "    #     Parameters:\n",
    "    #     - sequence: A string containing the amino acid sequence.\n",
    "    #     - header: A string to be used as the header in the FASTA file.\n",
    "    #     - file_name: The name of the FASTA file to be created.\n",
    "    #     \"\"\"\n",
    "    #     print(f\"Writing sequence to {file_name}\")\n",
    "    #     with open(file_name, 'w') as fasta_file:\n",
    "    #         # Write the header with the '>' symbol\n",
    "    #         fasta_file.write(f\">{header}\\n\")\n",
    "            \n",
    "    #         # Write the sequence in lines of 80 characters\n",
    "    #         for i in range(0, len(sequence), 80):\n",
    "    #             fasta_file.write(sequence[i:i+80] + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# fasta_path = os.path.join(\"raw_data\", \"MBP\", 'MBP_wt.fasta')\n",
    "# write_fasta(sequence, 'MBP_wt', fasta_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw_hdx_path = \"/Users/alexi/Library/CloudStorage/OneDrive-Nexus365/Rotation_Projects/Rotation_3/Project/ValDX/raw_data/MBP/MaltoseBindingProtein/MBP analysis final editing export 2020 10 05_tidy.csv\"\n",
    "# raw_hdx = pd.read_csv(raw_hdx_path)\n",
    "# raw_hdx.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# state = \"WT Null\"\n",
    "\n",
    "# raw_hdx = raw_hdx[raw_hdx['hx_sample'] == state]\n",
    "\n",
    "# # drop nans in column d\n",
    "# raw_hdx = raw_hdx.dropna(subset=['d'])\n",
    "\n",
    "# raw_hdx.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # group by pep_start and pep_end and hx_time and take the mean of the d values\n",
    "# grouped = raw_hdx.groupby(['pep_start', 'pep_end',\"hx_time\"])[\"d\"].mean().reset_index()\n",
    "\n",
    "# grouped.head()\n",
    "\n",
    "# # assign peptide number to each combination of pep_start and pep_end\n",
    "# grouped['peptide']= grouped.groupby(['pep_start', 'pep_end']).ngroup()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # print entire dataframe\n",
    "# pd.set_option('display.max_rows', None)\n",
    "# pd.set_option('display.max_columns', None)\n",
    "# pd.set_option('display.width', None)\n",
    "\n",
    "# print(grouped)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add MaxUptake column for each peptide\n",
    "\n",
    "# # first, get the max uptake for each peptide\n",
    "# max_uptake = grouped.groupby('peptide')['d'].max().reset_index()\n",
    "\n",
    "# # assign the max uptake to each peptide\n",
    "# grouped = grouped.merge(max_uptake, on='peptide', suffixes=('', '_max'))\n",
    "\n",
    "# grouped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grouped['UptakeFraction'] = grouped['d'] / grouped['d_max']\n",
    "\n",
    "\n",
    "\n",
    "# columns_to_drop = ['d', 'd_max']\n",
    "# grouped = grouped.drop(columns=columns_to_drop)\n",
    "\n",
    "# grouped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # pivot exposure and uptake fraction\n",
    "# grouped = grouped.pivot(index=['pep_start', 'pep_end'], columns='hx_time', values='UptakeFraction').reset_index()\n",
    "\n",
    "# # drop \n",
    "# grouped.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # print entire dataframe\n",
    "# pd.set_option('display.max_rows', None)\n",
    "# pd.set_option('display.max_columns', None)\n",
    "# pd.set_option('display.width', None)\n",
    "# print(grouped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # conver to HDXer format ie start, end, exposure_1, exposure_2 \n",
    "\n",
    "# # change Start to ResStr and End to ResEnd\n",
    "# hdx = grouped.rename(columns={'pep_start': 'ResStr', 'pep_end': 'ResEnd'})\n",
    "\n",
    "# # drop the exposure column\n",
    "# hdx.columns.name = None\n",
    "\n",
    "# print(hdx)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# hdx = hdx.round(5)\n",
    "# hdx.to_csv(os.path.join(\"raw_data\", \"MBP\", 'MBP_wt1.dat'), sep=' ', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# segs = hdx[['ResStr', 'ResEnd']].drop_duplicates().sort_values(by=['ResStr', 'ResEnd']).reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # convert to list of tuples\n",
    "# segs = [tuple(x) for x in segs.values]\n",
    "\n",
    "# print(segs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# # write list as new lines with space delimiter\n",
    "# with open(os.path.join(\"raw_data\", \"MBP\", 'MBP_wt1_segs.txt'), 'w') as f:\n",
    "#     for item in segs:\n",
    "#         f.write(\"%s\\n\" % ' '.join(map(str, item)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# BPTI_dir = \"/Users/alexi/Library/CloudStorage/OneDrive-Nexus365/Rotation_Projects/Rotation_3/Project/ValDX/raw_data/MBP/MaltoseBindingProtein\"\n",
    "# sim_dir = os.path.join(BPTI_dir, \"alphafold_quick\")\n",
    "\n",
    "# pdb_list = [f for f in os.listdir(sim_dir) if f.endswith('.pdb')]\n",
    "\n",
    "# print(pdb_list) \n",
    "\n",
    "\n",
    "# H_sim_dir = os.path.join(BPTI_dir, \"alphafold_H\")\n",
    "\n",
    "# os.makedirs(H_sim_dir, exist_ok=True)\n",
    "\n",
    "# for pdb in pdb_list:\n",
    "#     fixer = PDBFixer(os.path.join(sim_dir, pdb))\n",
    "#     fixer.addMissingHydrogens(7.0)\n",
    "#     H_pdb_name = pdb.replace('.pdb', '_H.pdb')\n",
    "#     PDBFile.writeFile(fixer.topology, fixer.positions, open(os.path.join(H_sim_dir, H_pdb_name), 'w'), keepIds=True)\n",
    "\n",
    "# pdb_list = [f for f in os.listdir(H_sim_dir) if f.endswith('.pdb')]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate conformations with Alphafold\n",
    "\n",
    "# need to find out how to generate a wide range of conformations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pre_process_main():\n",
    "    # BPTI data\n",
    "    BPTI_dir = \"/home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein\"\n",
    "\n",
    "    # BPTI_dir = \"/home/alexi/Documents/ValDX/raw_data/HDXer_tutorial/BPTI\"\n",
    "\n",
    "    sim_name = 'MBPwt_AF'\n",
    "    os.listdir(BPTI_dir)\n",
    "\n",
    "    segs_name = \"MBP_wt1_segs.txt\"\n",
    "    segs_path = os.path.join(BPTI_dir, segs_name)\n",
    "\n",
    "    hdx_name = \"MBP_wt1.dat\"\n",
    "    hdx_path = os.path.join(BPTI_dir, hdx_name)\n",
    "    print(hdx_path)\n",
    "\n",
    "    rates_name = \"out__train_MD_Simulated_1Intrinsic_rates.dat\"\n",
    "    rates_path = os.path.join(BPTI_dir, rates_name)\n",
    "\n",
    "    sim_dir = os.path.join(BPTI_dir, \"alphafold_quick\")\n",
    "\n",
    "    pdb_list = [f for f in os.listdir(sim_dir) if f.endswith('.pdb')]\n",
    "\n",
    "    print(pdb_list) \n",
    "\n",
    "\n",
    "    H_sim_dir = os.path.join(BPTI_dir, \"alphafold_H\")\n",
    "\n",
    "    os.makedirs(H_sim_dir, exist_ok=True)\n",
    "\n",
    "    for pdb in pdb_list:\n",
    "        fixer = PDBFixer(os.path.join(sim_dir, pdb))\n",
    "        fixer.addMissingHydrogens(7.0)\n",
    "        H_pdb_name = pdb.replace('.pdb', '_H.pdb')\n",
    "        PDBFile.writeFile(fixer.topology, fixer.positions, open(os.path.join(H_sim_dir, H_pdb_name), 'w'), keepIds=True)\n",
    "\n",
    "    pdb_list = [f for f in os.listdir(H_sim_dir) if f.endswith('.pdb')]\n",
    "\n",
    "\n",
    "    top_path = os.path.join(H_sim_dir, pdb_list[0])\n",
    "    pdb_paths = [os.path.join(H_sim_dir, i) for i in pdb_list]\n",
    "\n",
    "    print(top_path)\n",
    "    print(pdb_paths)\n",
    "\n",
    "\n",
    "    small_traj_name = top_path.replace(\".pdb\",\"_small.xtc\")\n",
    "    small_traj_path = os.path.join(sim_dir, small_traj_name)\n",
    "\n",
    "    u = mda.Universe(top_path, pdb_paths)\n",
    "\n",
    "\n",
    "        \n",
    "    with XTCWriter(small_traj_path, n_atoms=u.atoms.n_atoms) as W:\n",
    "        for ts in u.trajectory:\n",
    "                W.write(u.atoms)\n",
    "\n",
    "    # traj_paths = [os.path.join(sim_dir, i) for i in os.listdir(sim_dir) if i.endswith(\".pdb\")]\n",
    "    \n",
    "    traj_paths = [small_traj_path]\n",
    "\n",
    "    print(traj_paths)\n",
    "    return hdx_path, segs_path, rates_path, top_path, traj_paths, sim_name, expt_name, test_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m hdx_path, segs_path, rates_path, top_path, traj_paths, sim_name, expt_name, test_name \u001b[38;5;241m=\u001b[39m \u001b[43mpre_process_main\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[18], line 8\u001b[0m, in \u001b[0;36mpre_process_main\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# BPTI_dir = \"/home/alexi/Documents/ValDX/raw_data/HDXer_tutorial/BPTI\"\u001b[39;00m\n\u001b[1;32m      7\u001b[0m sim_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMBPwt_AF\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 8\u001b[0m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlistdir\u001b[49m\u001b[43m(\u001b[49m\u001b[43mBPTI_dir\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m segs_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMBP_wt1_segs.txt\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     11\u001b[0m segs_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(BPTI_dir, segs_name)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein'"
     ]
    }
   ],
   "source": [
    "hdx_path, segs_path, rates_path, top_path, traj_paths, sim_name, expt_name, test_name = pre_process_main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # BPTI data\n",
    "# BPTI_dir = \"/Users/alexi/Library/CloudStorage/OneDrive-Nexus365/Rotation_Projects/Rotation_3/Project/ValDX/raw_data/HDXer_tutorial/BPTI\"\n",
    "# # BPTI_dir = \"/home/alexi/Documents/ValDX/raw_data/HDXer_tutorial/BPTI\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# expt_dir = os.path.join(BPTI_dir, \"BPTI_expt_data\")\n",
    "\n",
    "# os.listdir(expt_dir)\n",
    "\n",
    "# segs_name = \"BPTI_residue_segs.txt\"\n",
    "# segs_path = os.path.join(expt_dir, segs_name)\n",
    "\n",
    "# hdx_name = \"BPTI_expt_dfracs.dat\"\n",
    "# hdx_path = os.path.join(expt_dir, hdx_name)\n",
    "# print(hdx_path)\n",
    "\n",
    "# rates_name = \"BPTI_Intrinsic_rates.dat\"\n",
    "# rates_path = os.path.join(expt_dir, rates_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sim_name = 'BPTI_MD'\n",
    "\n",
    "# sim_dir = os.path.join(BPTI_dir, \"BPTI_simulations\")\n",
    "\n",
    "# os.listdir(sim_dir)\n",
    "\n",
    "# md_reps = 1\n",
    "# rep_dirs = [\"Run_\"+str(i+1) for i in range(md_reps)]\n",
    "\n",
    "# top_name = \"bpti_5pti_eq6_protonly.gro\"\n",
    "\n",
    "# top_path = os.path.join(sim_dir, rep_dirs[0], top_name)\n",
    "\n",
    "# traj_name = \"bpti_5pti_reimg_protonly.xtc\"\n",
    "\n",
    "# traj_paths = [os.path.join(sim_dir, rep_dir, traj_name) for rep_dir in rep_dirs]\n",
    "\n",
    "# print(top_path)\n",
    "# print(traj_paths)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_split_test(split_mode, name, system):\n",
    "\n",
    "    # settings.split_mode = 'R'\n",
    "    settings.split_mode = split_mode\n",
    "    settings.name = \"_\".join([name, split_mode])\n",
    "    settings.times = [30, 240, 1800, 14400,]\n",
    "    VDX = ValDXer(settings)\n",
    "\n",
    "    VDX.load_HDX_data(HDX_path=hdx_path, SEG_path=segs_path, calc_name=expt_name)\n",
    "    # VDX.load_intrinsic_rates(path=rates_path, calc_name=expt_name)\n",
    "\n",
    "    VDX.load_structures(top_path=top_path, traj_paths=traj_paths, calc_name=test_name)\n",
    "\n",
    "    run_outputs = VDX.run_VDX(calc_name=test_name, expt_name=expt_name)\n",
    "    analysis_dump, df, name = VDX.dump_analysis()\n",
    "    save_path = VDX.save_experiment()\n",
    "\n",
    "    return run_outputs, analysis_dump, df, name, save_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "envs ['# conda environments:', '#', 'base                     /home/alexi/anaconda3', 'HDXER_ENV             *  /home/alexi/anaconda3/envs/HDXER_ENV', 'openfold_env             /home/alexi/anaconda3/envs/openfold_env', 'plumed-masterclass-2022     /home/alexi/anaconda3/envs/plumed-masterclass-2022', '                         /home/alexi/localcolabfold/colabfold-conda', '                         /home/alexi/localcolabfold/conda', '']\n",
      "name base\n",
      "name HDXER_ENV\n",
      "Environment 'HDXER_ENV' found.\n",
      "Path to 'HDXER_ENV' environment: /home/alexi/anaconda3/envs/HDXER_ENV\n",
      "PATH /home/alexi/anaconda3/envs/HDXER_ENV:/home/alexi/anaconda3/envs/HDXER_ENV/bin:/home/alexi/anaconda3/condabin:/home/alexi/.vscode/cli/servers/Stable-903b1e9d8990623e3d7da1df3d33db3e42d80eda/server/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin\n",
      "calc_hdx\n",
      "/home/alexi/Documents/HDXer/HDXer/calc_hdx.py\n",
      "Preparing HDX data for Experimental\n",
      "                                                 HDX  \\\n",
      "0  /home/alexi/Documents/ValDX/raw_data/MBP/Malto...   \n",
      "\n",
      "                                                 SEG     calc_name  \\\n",
      "0  /home/alexi/Documents/ValDX/raw_data/MBP/Malto...  Experimental   \n",
      "\n",
      "   experimental  \n",
      "0         False  \n",
      "/home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein/MBP_wt1.dat\n",
      "Path /home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein/MBP_wt1.dat\n",
      "AVG: ncol = 6, len(names) = 4\n",
      "                                                 HDX  \\\n",
      "0  /home/alexi/Documents/ValDX/raw_data/MBP/Malto...   \n",
      "1                                                NaN   \n",
      "\n",
      "                                                 SEG     calc_name  \\\n",
      "0  /home/alexi/Documents/ValDX/raw_data/MBP/Malto...  Experimental   \n",
      "1                                                NaN   MBPwt1_test   \n",
      "\n",
      "  experimental                                                top  \\\n",
      "0        False                                                NaN   \n",
      "1          NaN  /home/alexi/Documents/ValDX/raw_data/MBP/Malto...   \n",
      "\n",
      "                                                traj  \n",
      "0                                                NaN  \n",
      "1  [/home/alexi/Documents/ValDX/raw_data/MBP/Malt...  \n",
      "Structures loaded MBPwt1_test: \n",
      "MBPwt1_test Topology: <Universe with 6131 atoms>\n",
      "MBPwt1_test Trajectory: <Universe with 6131 atoms>\n",
      "MBPwt1_test Traj: no frames 5\n",
      "Running VDX loop\n",
      "Random seeds: [42, 43, 44]\n",
      "Splitting segments for MBPwt1_test by spatial split: PCA1D\n",
      "HDX residues:  [ 19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36\n",
      "  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54\n",
      "  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72\n",
      "  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90\n",
      "  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108\n",
      " 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126\n",
      " 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144\n",
      " 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 162 163\n",
      " 164 165 166 167 168 173 174 175 176 177 178 179 180 181 182 183 184 185\n",
      " 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203\n",
      " 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221\n",
      " 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239\n",
      " 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257\n",
      " 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275\n",
      " 276 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293\n",
      " 294 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311\n",
      " 312 313 314 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329\n",
      " 330 331 332 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347\n",
      " 348 349 350 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365\n",
      " 366 367 368 369 370 371 372 373 374 375 376 377]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexi/anaconda3/envs/HDXER_ENV/lib/python3.8/site-packages/MDAnalysis/lib/mdamath.py:300: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  alpha = np.rad2deg(np.arccos(np.dot(y, z) / (ly * lz)))\n",
      "/home/alexi/anaconda3/envs/HDXER_ENV/lib/python3.8/site-packages/MDAnalysis/lib/mdamath.py:301: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  beta = np.rad2deg(np.arccos(np.dot(x, z) / (lx * lz)))\n",
      "/home/alexi/anaconda3/envs/HDXER_ENV/lib/python3.8/site-packages/MDAnalysis/lib/mdamath.py:302: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  gamma = np.rad2deg(np.arccos(np.dot(x, y) / (lx * ly)))\n",
      "/home/alexi/Documents/ValDX/ValDX/Experiment_ABC.py:735: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  train_segs[\"calc_name\"] = train_rep_name\n",
      "/home/alexi/Documents/ValDX/ValDX/Experiment_ABC.py:736: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  val_segs[\"calc_name\"] = val_rep_name\n",
      "/home/alexi/Documents/ValDX/ValDX/Experiment_ABC.py:750: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  train_segs[\"path\"] = [train_segs_path]*len(train_segs)\n",
      "/home/alexi/Documents/ValDX/ValDX/Experiment_ABC.py:751: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  val_segs[\"path\"] = [val_segs_path]*len(val_segs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train_segs\n",
      "    ResStr  ResEnd  peptide            calc_name\n",
      "35     140     146       35  train_MBPwt1_test_1\n",
      "36     147     158       36  train_MBPwt1_test_1\n",
      "37     147     159       37  train_MBPwt1_test_1\n",
      "38     147     160       38  train_MBPwt1_test_1\n",
      "39     162     168       39  train_MBPwt1_test_1\n",
      "Saved train MBPwt1_test_1 segments to /home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1/train_residue_segs__MBPwt1_test_.txt\n",
      "Train Peptide numbers: [ 35  36  37  38  39  40  41  42  43  44  45  51  52  53  54  55  56  57\n",
      "  58  59  60  61  62  63  64  65  66  67  99 100 101 102 104 105 114]\n",
      "Saved val MBPwt1_test_1 segments to /home/alexi/Documents/ValDX/data/AvsB_S/val_MBPwt1_test_1/val_residue_segs__MBPwt1_test_.txt\n",
      "Val Peptide numbers: [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17\n",
      "  18  19  20  21  22  23  24  29  34  78  79  80  81  82  83  84  85  86\n",
      "  87  88  89  90  91  92 109]\n",
      "train_HDX_data\n",
      "         30      240     1800    14400  peptide            calc_name  \\\n",
      "0   0.35002  0.74064  0.93911  1.00000       35  train_MBPwt1_test_1   \n",
      "1   0.28447  0.44235  0.62680  1.00000       36  train_MBPwt1_test_1   \n",
      "2   0.27671  0.42953  0.60748  1.00000       37  train_MBPwt1_test_1   \n",
      "3   0.27058  0.44360  0.63441  1.00000       38  train_MBPwt1_test_1   \n",
      "4   0.09577  0.11538  0.29052  1.00000       39  train_MBPwt1_test_1   \n",
      "5   0.08464  0.08493  0.25458  1.00000       40  train_MBPwt1_test_1   \n",
      "6   0.46218  0.55885  0.74861  1.00000       41  train_MBPwt1_test_1   \n",
      "7   0.91123  1.00000  0.99020  0.99646       42  train_MBPwt1_test_1   \n",
      "8   0.94532  1.00000  0.97612  0.99286       43  train_MBPwt1_test_1   \n",
      "9   0.94076  1.00000  0.99297  0.99856       44  train_MBPwt1_test_1   \n",
      "10  0.92347  1.00000  0.98426  0.99716       45  train_MBPwt1_test_1   \n",
      "11  0.06773  0.05906  0.23257  1.00000       51  train_MBPwt1_test_1   \n",
      "12  0.14384  0.13577  0.28317  1.00000       52  train_MBPwt1_test_1   \n",
      "13  0.47719  0.59213  0.81960  1.00000       53  train_MBPwt1_test_1   \n",
      "14  0.50658  0.62809  0.81196  1.00000       54  train_MBPwt1_test_1   \n",
      "15  0.47039  0.57622  0.82908  1.00000       55  train_MBPwt1_test_1   \n",
      "16  0.51577  0.62619  0.83942  1.00000       56  train_MBPwt1_test_1   \n",
      "17  0.55840  0.67967  0.86835  1.00000       57  train_MBPwt1_test_1   \n",
      "18  0.30407  0.49791  0.67604  1.00000       58  train_MBPwt1_test_1   \n",
      "19  0.06351  0.10488  0.24644  1.00000       59  train_MBPwt1_test_1   \n",
      "20  0.19954  0.21076  0.36065  1.00000       60  train_MBPwt1_test_1   \n",
      "21  0.18095  0.20552  0.37845  1.00000       61  train_MBPwt1_test_1   \n",
      "22  0.20376  0.22998  0.38026  1.00000       62  train_MBPwt1_test_1   \n",
      "23  0.19713  0.20063  0.39872  1.00000       63  train_MBPwt1_test_1   \n",
      "24  0.20340  0.20803  0.39135  1.00000       64  train_MBPwt1_test_1   \n",
      "25  0.22350  0.23010  0.41531  1.00000       65  train_MBPwt1_test_1   \n",
      "26  0.48976  0.56316  0.70966  1.00000       66  train_MBPwt1_test_1   \n",
      "27  0.53547  0.71996  0.86150  1.00000       67  train_MBPwt1_test_1   \n",
      "28  0.56266  0.73956  0.85879  1.00000       99  train_MBPwt1_test_1   \n",
      "29  0.64179  0.78590  0.88306  1.00000      100  train_MBPwt1_test_1   \n",
      "30  0.68481  0.82015  0.90430  1.00000      101  train_MBPwt1_test_1   \n",
      "31  0.89492  1.00000  0.98833  0.99306      102  train_MBPwt1_test_1   \n",
      "32  0.29123  0.73652  0.96097  1.00000      104  train_MBPwt1_test_1   \n",
      "33  0.31020  0.71462  0.94659  1.00000      105  train_MBPwt1_test_1   \n",
      "34  0.91682  1.00000  0.98034  0.98742      114  train_MBPwt1_test_1   \n",
      "\n",
      "                                                 path  ResStr  ResEnd  \n",
      "0   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     140     146  \n",
      "1   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     147     158  \n",
      "2   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     147     159  \n",
      "3   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     147     160  \n",
      "4   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     162     168  \n",
      "5   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     163     168  \n",
      "6   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     173     187  \n",
      "7   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     174     179  \n",
      "8   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     178     186  \n",
      "9   /home/alexi/Documents/ValDX/data/AvsB_S/train_...     180     187  \n",
      "10  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     180     191  \n",
      "11  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     204     206  \n",
      "12  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     206     209  \n",
      "13  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     206     221  \n",
      "14  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     207     217  \n",
      "15  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     207     220  \n",
      "16  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     207     221  \n",
      "17  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     209     221  \n",
      "18  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     221     225  \n",
      "19  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     222     227  \n",
      "20  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     225     235  \n",
      "21  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     226     234  \n",
      "22  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     226     235  \n",
      "23  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     228     234  \n",
      "24  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     228     235  \n",
      "25  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     229     235  \n",
      "26  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     236     245  \n",
      "27  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     246     253  \n",
      "28  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     334     347  \n",
      "29  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     334     349  \n",
      "30  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     334     350  \n",
      "31  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     342     350  \n",
      "32  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     350     356  \n",
      "33  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     351     356  \n",
      "34  /home/alexi/Documents/ValDX/data/AvsB_S/train_...     370     377  \n",
      "    ResStr  ResEnd       30      240     1800    14400  peptide  \\\n",
      "0      140     146  0.35002  0.74064  0.93911  1.00000       35   \n",
      "1      147     158  0.28447  0.44235  0.62680  1.00000       36   \n",
      "2      147     159  0.27671  0.42953  0.60748  1.00000       37   \n",
      "3      147     160  0.27058  0.44360  0.63441  1.00000       38   \n",
      "4      162     168  0.09577  0.11538  0.29052  1.00000       39   \n",
      "5      163     168  0.08464  0.08493  0.25458  1.00000       40   \n",
      "6      173     187  0.46218  0.55885  0.74861  1.00000       41   \n",
      "7      174     179  0.91123  1.00000  0.99020  0.99646       42   \n",
      "8      178     186  0.94532  1.00000  0.97612  0.99286       43   \n",
      "9      180     187  0.94076  1.00000  0.99297  0.99856       44   \n",
      "10     180     191  0.92347  1.00000  0.98426  0.99716       45   \n",
      "11     204     206  0.06773  0.05906  0.23257  1.00000       51   \n",
      "12     206     209  0.14384  0.13577  0.28317  1.00000       52   \n",
      "13     206     221  0.47719  0.59213  0.81960  1.00000       53   \n",
      "14     207     217  0.50658  0.62809  0.81196  1.00000       54   \n",
      "15     207     220  0.47039  0.57622  0.82908  1.00000       55   \n",
      "16     207     221  0.51577  0.62619  0.83942  1.00000       56   \n",
      "17     209     221  0.55840  0.67967  0.86835  1.00000       57   \n",
      "18     221     225  0.30407  0.49791  0.67604  1.00000       58   \n",
      "19     222     227  0.06351  0.10488  0.24644  1.00000       59   \n",
      "20     225     235  0.19954  0.21076  0.36065  1.00000       60   \n",
      "21     226     234  0.18095  0.20552  0.37845  1.00000       61   \n",
      "22     226     235  0.20376  0.22998  0.38026  1.00000       62   \n",
      "23     228     234  0.19713  0.20063  0.39872  1.00000       63   \n",
      "24     228     235  0.20340  0.20803  0.39135  1.00000       64   \n",
      "25     229     235  0.22350  0.23010  0.41531  1.00000       65   \n",
      "26     236     245  0.48976  0.56316  0.70966  1.00000       66   \n",
      "27     246     253  0.53547  0.71996  0.86150  1.00000       67   \n",
      "28     334     347  0.56266  0.73956  0.85879  1.00000       99   \n",
      "29     334     349  0.64179  0.78590  0.88306  1.00000      100   \n",
      "30     334     350  0.68481  0.82015  0.90430  1.00000      101   \n",
      "31     342     350  0.89492  1.00000  0.98833  0.99306      102   \n",
      "32     350     356  0.29123  0.73652  0.96097  1.00000      104   \n",
      "33     351     356  0.31020  0.71462  0.94659  1.00000      105   \n",
      "34     370     377  0.91682  1.00000  0.98034  0.98742      114   \n",
      "\n",
      "              calc_name                                               path  \n",
      "0   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "1   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "2   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "3   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "4   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "5   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "6   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "7   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "8   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "9   train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "10  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "11  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "12  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "13  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "14  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "15  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "16  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "17  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "18  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "19  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "20  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "21  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "22  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "23  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "24  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "25  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "26  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "27  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "28  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "29  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "30  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "31  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "32  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "33  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "34  train_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/train_...  \n",
      "#\tResStr\tResEnd\t30\t240\t1800\t14400\t times/min\n",
      "\n",
      "Saved train MBPwt1_test HDX data to /home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1/train_MBPwt1_test_1_expt_dfracs.dat\n",
      "    ResStr  ResEnd       30      240     1800    14400  peptide  \\\n",
      "0       19      30  0.39411  0.66768  0.90249  1.00000        0   \n",
      "1       19      31  0.37620  0.63567  0.86921  1.00000        1   \n",
      "2       24      35  0.15136  0.22110  0.46211  1.00000        2   \n",
      "3       31      53  0.36285  0.47337  0.69324  1.00000        3   \n",
      "4       32      46  0.12702  0.19595  0.46017  1.00000        4   \n",
      "5       47      54  0.64204  0.79488  0.96913  1.00000        5   \n",
      "6       55      72  0.43593  0.58470  0.83613  1.00000        6   \n",
      "7       73      80  0.57720  0.80352  0.95520  1.00000        7   \n",
      "8       73      81  0.57986  0.83787  0.96932  1.00000        8   \n",
      "9       73      86  0.53644  0.81494  0.92973  1.00000        9   \n",
      "10      73      87  0.47594  0.72899  0.87130  1.00000       10   \n",
      "11      81      86  0.51513  0.78972  0.88611  1.00000       11   \n",
      "12      81      87  0.41829  0.63869  0.76793  1.00000       12   \n",
      "13      82      87  0.39984  0.55747  0.72764  1.00000       13   \n",
      "14      87     100  0.54873  0.67583  0.80366  1.00000       14   \n",
      "15      87     104  0.59452  0.71314  0.82909  1.00000       15   \n",
      "16      87     106  0.57272  0.68998  0.81656  1.00000       16   \n",
      "17      87     107  0.58395  0.69213  0.81831  1.00000       17   \n",
      "18      88     100  0.56700  0.64136  0.78843  1.00000       18   \n",
      "19      88     106  0.57519  0.66131  0.80426  1.00000       19   \n",
      "20      88     107  0.57756  0.66306  0.80102  1.00000       20   \n",
      "21      96     106  0.58892  0.64734  0.71765  1.00000       21   \n",
      "22     105     114  0.55484  0.72145  0.79427  1.00000       22   \n",
      "23     106     114  0.56888  0.72457  0.80402  1.00000       23   \n",
      "24     108     114  0.52872  0.71398  0.80058  1.00000       24   \n",
      "25     122     126  0.11896  0.15998  0.29444  1.00000       29   \n",
      "26     130     133  0.38101  0.91543  0.96103  1.00000       34   \n",
      "27     290     295  1.00000  0.19126  0.44809  0.65574       78   \n",
      "28     294     301  0.41170  0.65760  0.88686  1.00000       79   \n",
      "29     294     312  0.20894  0.34105  0.57924  1.00000       80   \n",
      "30     295     300  0.40234  0.67281  0.86057  1.00000       81   \n",
      "31     296     302  0.42179  0.68613  0.85949  1.00000       82   \n",
      "32     296     303  0.44354  0.67838  0.83253  1.00000       83   \n",
      "33     296     312  0.17884  0.27933  0.52371  1.00000       84   \n",
      "34     301     314  0.15684  0.22064  0.46063  1.00000       85   \n",
      "35     302     312  0.14446  0.21221  0.46010  1.00000       86   \n",
      "36     303     312  0.13635  0.20690  0.45974  1.00000       87   \n",
      "37     304     312  0.13211  0.20321  0.45953  1.00000       88   \n",
      "38     313     317  0.80239  0.96967  0.96573  1.00000       89   \n",
      "39     313     319  0.39720  0.48589  0.57606  1.00000       90   \n",
      "40     313     321  0.28690  0.35552  0.45958  1.00000       91   \n",
      "41     313     322  0.25978  0.33700  0.50240  1.00000       92   \n",
      "42     357     361  0.59870  0.85358  0.98298  1.00000      109   \n",
      "\n",
      "            calc_name                                               path  \n",
      "0   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "1   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "2   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "3   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "4   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "5   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "6   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "7   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "8   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "9   val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "10  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "11  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "12  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "13  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "14  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "15  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "16  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "17  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "18  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "19  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "20  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "21  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "22  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "23  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "24  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "25  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "26  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "27  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "28  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "29  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "30  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "31  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "32  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "33  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "34  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "35  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "36  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "37  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "38  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "39  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "40  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "41  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "42  val_MBPwt1_test_1  /home/alexi/Documents/ValDX/data/AvsB_S/val_MB...  \n",
      "#\tResStr\tResEnd\t30\t240\t1800\t14400\t times/min\n",
      "\n",
      "Saved val MBPwt1_test HDX data to /home/alexi/Documents/ValDX/data/AvsB_S/val_MBPwt1_test_1/val_MBPwt1_test_1_expt_dfracs.dat\n",
      "/home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1\n",
      "MBPwt1_test\n",
      "/home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1/out__train_MBPwt1_test_1\n",
      "envs ['# conda environments:', '#', 'base                     /home/alexi/anaconda3', 'HDXER_ENV             *  /home/alexi/anaconda3/envs/HDXER_ENV', 'openfold_env             /home/alexi/anaconda3/envs/openfold_env', 'plumed-masterclass-2022     /home/alexi/anaconda3/envs/plumed-masterclass-2022', '                         /home/alexi/localcolabfold/colabfold-conda', '                         /home/alexi/localcolabfold/conda', '']\n",
      "name base\n",
      "name HDXER_ENV\n",
      "Environment 'HDXER_ENV' found.\n",
      "Path to 'HDXER_ENV' environment: /home/alexi/anaconda3/envs/HDXER_ENV\n",
      "PATH /home/alexi/anaconda3/envs/HDXER_ENV:/home/alexi/anaconda3/envs/HDXER_ENV/bin:/home/alexi/anaconda3/condabin:/home/alexi/.vscode/cli/servers/Stable-903b1e9d8990623e3d7da1df3d33db3e42d80eda/server/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin\n",
      "calc_hdx\n",
      "/home/alexi/Documents/HDXer/HDXer/calc_hdx.py\n",
      "conda run -n HDXER_ENV python /home/alexi/Documents/HDXer/HDXer/calc_hdx.py -t /home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein/alphafold_H/MBP_wt_unrelaxed_rank_004_alphafold2_ptm_model_2_seed_000_H_small.xtc -p /home/alexi/Documents/ValDX/raw_data/MBP/MaltoseBindingProtein/alphafold_H/MBP_wt_unrelaxed_rank_004_alphafold2_ptm_model_2_seed_000_H.pdb -m BestVendruscolo -log /home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1/calc_hdx_train_MBPwt1_test_1.log -out /home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1/out__train_MBPwt1_test_1 -seg /home/alexi/Documents/ValDX/data/AvsB_S/train_MBPwt1_test_1/train_residue_segs__MBPwt1_test_.txt -mopt \"{ 'save_detailed' : True }\" --times 30 240 1800 14400 -str 1\n",
      "envs ['# conda environments:', '#', 'base                     /home/alexi/anaconda3', 'HDXER_ENV             *  /home/alexi/anaconda3/envs/HDXER_ENV', 'openfold_env             /home/alexi/anaconda3/envs/openfold_env', 'plumed-masterclass-2022     /home/alexi/anaconda3/envs/plumed-masterclass-2022', '                         /home/alexi/localcolabfold/colabfold-conda', '                         /home/alexi/localcolabfold/conda', '']\n",
      "name base\n",
      "name HDXER_ENV\n",
      "Environment 'HDXER_ENV' found.\n",
      "Path to 'HDXER_ENV' environment: /home/alexi/anaconda3/envs/HDXER_ENV\n",
      "PATH /home/alexi/anaconda3/envs/HDXER_ENV:/home/alexi/anaconda3/envs/HDXER_ENV/bin:/home/alexi/anaconda3/condabin:/home/alexi/.vscode/cli/servers/Stable-903b1e9d8990623e3d7da1df3d33db3e42d80eda/server/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 13\u001b[0m\n\u001b[1;32m      9\u001b[0m save_paths \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m split, split_name \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(splits, split_names):\n\u001b[0;32m---> 13\u001b[0m     run_outputs, analysis_dump, df, name, save_path \u001b[38;5;241m=\u001b[39m \u001b[43mrun_split_test\u001b[49m\u001b[43m(\u001b[49m\u001b[43msplit\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msplit_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msystem\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     14\u001b[0m     raw_run_outputs[name] \u001b[38;5;241m=\u001b[39m run_outputs\n\u001b[1;32m     15\u001b[0m     analysis_dumps\u001b[38;5;241m.\u001b[39mupdate(analysis_dump)\n",
      "Cell \u001b[0;32mIn[27], line 14\u001b[0m, in \u001b[0;36mrun_split_test\u001b[0;34m(split_mode, name, system)\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# VDX.load_intrinsic_rates(path=rates_path, calc_name=expt_name)\u001b[39;00m\n\u001b[1;32m     12\u001b[0m VDX\u001b[38;5;241m.\u001b[39mload_structures(top_path\u001b[38;5;241m=\u001b[39mtop_path, traj_paths\u001b[38;5;241m=\u001b[39mtraj_paths, calc_name\u001b[38;5;241m=\u001b[39mtest_name)\n\u001b[0;32m---> 14\u001b[0m run_outputs \u001b[38;5;241m=\u001b[39m \u001b[43mVDX\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_VDX\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcalc_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtest_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexpt_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpt_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     15\u001b[0m analysis_dump, df, name \u001b[38;5;241m=\u001b[39m VDX\u001b[38;5;241m.\u001b[39mdump_analysis()\n\u001b[1;32m     16\u001b[0m save_path \u001b[38;5;241m=\u001b[39m VDX\u001b[38;5;241m.\u001b[39msave_experiment()\n",
      "File \u001b[0;32m~/Documents/ValDX/ValDX/ValidationDX.py:635\u001b[0m, in \u001b[0;36mValDXer.run_VDX\u001b[0;34m(self, calc_name, expt_name, mode, n_reps, weights, random_seeds)\u001b[0m\n\u001b[1;32m    629\u001b[0m _, train_rep_name, val_rep_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msplit_segments(seg_name\u001b[38;5;241m=\u001b[39mexpt_name,\n\u001b[1;32m    630\u001b[0m                                                       calc_name\u001b[38;5;241m=\u001b[39mcalc_name, \n\u001b[1;32m    631\u001b[0m                                                       rep\u001b[38;5;241m=\u001b[39mrep, \n\u001b[1;32m    632\u001b[0m                                                       random_seed\u001b[38;5;241m=\u001b[39mrandom_seeds[rep\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n\u001b[1;32m    634\u001b[0m \u001b[38;5;66;03m# train HDX\u001b[39;00m\n\u001b[0;32m--> 635\u001b[0m train_opt_gamma, train_df, cr_bc_bh \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain_HDX\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcalc_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcalc_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    636\u001b[0m \u001b[43m                                           \u001b[49m\u001b[43mexpt_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpt_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    637\u001b[0m \u001b[43m                                           \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    638\u001b[0m \u001b[43m                                           \u001b[49m\u001b[43mweights\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweights\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    639\u001b[0m \u001b[43m                                           \u001b[49m\u001b[43mrep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrep\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    640\u001b[0m train_gammas\u001b[38;5;241m.\u001b[39mappend(train_opt_gamma)\n\u001b[1;32m    642\u001b[0m \u001b[38;5;66;03m# validation HDX\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/ValDX/ValDX/ValidationDX.py:1002\u001b[0m, in \u001b[0;36mValDXer.train_HDX\u001b[0;34m(self, calc_name, expt_name, mode, weights, rep)\u001b[0m\n\u001b[1;32m    990\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mtrain_HDX\u001b[39m(\u001b[38;5;28mself\u001b[39m, \n\u001b[1;32m    991\u001b[0m               calc_name: \u001b[38;5;28mstr\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \n\u001b[1;32m    992\u001b[0m               expt_name: \u001b[38;5;28mstr\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    999\u001b[0m \n\u001b[1;32m   1000\u001b[0m     \u001b[38;5;66;03m# for rep in range(n_reps):\u001b[39;00m\n\u001b[0;32m-> 1002\u001b[0m     rep_df, _ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict_HDX\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcalc_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcalc_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m   1003\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mrep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m   1004\u001b[0m \u001b[43m                                 \u001b[49m\u001b[43mtrain\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m   1006\u001b[0m     \u001b[38;5;66;03m# add df to HDX_data\u001b[39;00m\n\u001b[1;32m   1008\u001b[0m     gamma, df, cr_bc_bh \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreweight_HDX(expt_name\u001b[38;5;241m=\u001b[39mexpt_name, \n\u001b[1;32m   1009\u001b[0m                                   calc_name\u001b[38;5;241m=\u001b[39mcalc_name, \n\u001b[1;32m   1010\u001b[0m                                   train\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, \n\u001b[1;32m   1011\u001b[0m                                   weights\u001b[38;5;241m=\u001b[39mweights,\n\u001b[1;32m   1012\u001b[0m                                   rep\u001b[38;5;241m=\u001b[39mrep)\n",
      "File \u001b[0;32m~/Documents/ValDX/ValDX/ValidationDX.py:207\u001b[0m, in \u001b[0;36mValDXer.predict_HDX\u001b[0;34m(self, calc_name, rep, train)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[38;5;66;03m# print(\" \".join(calc_hdx_command))\u001b[39;00m\n\u001b[1;32m    205\u001b[0m env_path \u001b[38;5;241m=\u001b[39m conda_to_env_dict(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mHDXer_env)\n\u001b[0;32m--> 207\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcalc_hdx_command\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    208\u001b[0m \u001b[43m               \u001b[49m\u001b[43menv\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43menv_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    209\u001b[0m \u001b[43m               \u001b[49m\u001b[43mshell\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    210\u001b[0m \u001b[43m               \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    211\u001b[0m \u001b[43m               \u001b[49m\u001b[43mcwd\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout_dir\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;66;03m### read HDX data into df\u001b[39;00m\n\u001b[1;32m    215\u001b[0m df \u001b[38;5;241m=\u001b[39m dfracs_to_df(out_prefix \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSegment_average_fractions.dat\u001b[39m\u001b[38;5;124m\"\u001b[39m, \n\u001b[1;32m    216\u001b[0m                   names\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettings\u001b[38;5;241m.\u001b[39mtimes)\n",
      "File \u001b[0;32m~/anaconda3/envs/HDXER_ENV/lib/python3.8/subprocess.py:495\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[1;32m    494\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 495\u001b[0m         stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcommunicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    496\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TimeoutExpired \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    497\u001b[0m         process\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m~/anaconda3/envs/HDXER_ENV/lib/python3.8/subprocess.py:1020\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1018\u001b[0m         stderr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m   1019\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m-> 1020\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1021\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1022\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/HDXER_ENV/lib/python3.8/subprocess.py:1083\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1081\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m _time() \u001b[38;5;241m+\u001b[39m timeout\n\u001b[1;32m   1082\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1083\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1084\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1085\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1086\u001b[0m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1087\u001b[0m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1088\u001b[0m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1089\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/HDXER_ENV/lib/python3.8/subprocess.py:1822\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1820\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1821\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1822\u001b[0m (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1823\u001b[0m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1824\u001b[0m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1825\u001b[0m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1826\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pid \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid:\n",
      "File \u001b[0;32m~/anaconda3/envs/HDXER_ENV/lib/python3.8/subprocess.py:1780\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1778\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1779\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1780\u001b[0m     (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1781\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1782\u001b[0m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1784\u001b[0m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1785\u001b[0m     pid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "\n",
    "splits = ['S', ]\n",
    "split_names = ['AvsB']\n",
    "system = 'BPTITtut_test'\n",
    "\n",
    "raw_run_outputs = {}\n",
    "analysis_dumps = {}\n",
    "analysis_df = pd.DataFrame()\n",
    "names = []\n",
    "save_paths = []\n",
    "\n",
    "\n",
    "for split, split_name in zip(splits, split_names):\n",
    "    run_outputs, analysis_dump, df, name, save_path = run_split_test(split, split_name, system)\n",
    "    raw_run_outputs[name] = run_outputs\n",
    "    analysis_dumps.update(analysis_dump)\n",
    "    analysis_df = pd.concat([analysis_df, df])\n",
    "    names.append(name)\n",
    "    save_paths.append(save_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming your DataFrame is named df\n",
    "# Replace 'your_dataframe' with your actual DataFrame variable\n",
    "df = analysis_df\n",
    "\n",
    "# Create a FacetGrid, using 'name' for each subplot\n",
    "g = sns.FacetGrid(df, col=\"name\", col_wrap=3, height=4, aspect=1.5)\n",
    "g.fig.suptitle('MSE over Time by Type for each Named Split Mode')\n",
    "\n",
    "# Create boxplots\n",
    "g = g.map(sns.boxplot, \"time\", \"mse\", \"Type\", palette=\"Set3\")\n",
    "\n",
    "# Adding some additional options for better visualization\n",
    "g.add_legend(title='Type')\n",
    "g.set_axis_labels(\"Time\", \"MSE\")\n",
    "g.set_titles(\"{col_name}\")\n",
    "\n",
    "# Adjust the arrangement of the plots\n",
    "plt.subplots_adjust(top=0.9)\n",
    "\n",
    "# Show plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming your DataFrame is named df\n",
    "# Replace 'your_dataframe' with your actual DataFrame variable\n",
    "df = analysis_df\n",
    "\n",
    "# Create a FacetGrid, using 'name' for each subplot\n",
    "g = sns.FacetGrid(df, col=\"name\", col_wrap=3, height=4, aspect=1.5)\n",
    "g.fig.suptitle('R over Time by Type for each Named Split Mode')\n",
    "\n",
    "# Create boxplots\n",
    "g = g.map(sns.boxplot, \"time\", \"R\", \"Type\", palette=\"Set3\")\n",
    "\n",
    "# Adding some additional options for better visualization\n",
    "g.add_legend(title='Type')\n",
    "g.set_axis_labels(\"Time\", \"R\")\n",
    "g.set_titles(\"{col_name}\")\n",
    "\n",
    "# Adjust the arrangement of the plots\n",
    "plt.subplots_adjust(top=0.9)\n",
    "\n",
    "# Show plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot LogPfs by Residues colour by calc_name facet wrap by name\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "LogPfs = pd.concat([analysis_dumps[i][\"LogPfs\"] for i in names])\n",
    "\n",
    "print(LogPfs)\n",
    "\n",
    "LogPfs_df = LogPfs.explode(['LogPf','Residues'])\n",
    "\n",
    "\n",
    "# Create a FacetGrid, using 'name' for each subplot\n",
    "g = sns.FacetGrid(LogPfs_df, col=\"name\", col_wrap=3, height=4, aspect=1.5)\n",
    "g.fig.suptitle('LogPfs over Residues for each Named Split Mode')\n",
    "\n",
    "# Create lineplots\n",
    "g = g.map(sns.lineplot, \"Residues\", \"LogPf\", \"calc_name\", palette=\"Set2\")\n",
    "\n",
    "# Adding some additional options for better visualization\n",
    "g.add_legend(title='calc_name')\n",
    "g.set_axis_labels(\"Residues\", \"LogPf\")\n",
    "g.set_titles(\"{col_name}\")\n",
    "\n",
    "# Adjust the arrangement of the plots\n",
    "plt.subplots_adjust(top=0.9)\n",
    "\n",
    "# Show plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from MDAnalysis.analysis.dssp import DSSP\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VDX.paths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(top_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdb_test = mda.Universe(top_path)\n",
    "\n",
    "# # write out as a pdb and add header\n",
    "# pdb_test.atoms.write('test.pdb')\n",
    "# with open('test.pdb', 'r') as original: data = original.read()\n",
    "# with open('test.pdb', 'w') as modified: modified.write('HEADER    '+sim_name+'\\n'+data)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# def PDB_to_DSSP(top_path: str, dssp_path: str=None, sim_name: str=None):\n",
    "#     \"\"\"\n",
    "#     Run DSSP on a PDB file to generate a DSSP file. Reads the output and returns a list of secondary structure elements.\n",
    "#     Secondary structure elements are reduced to a single character: H (alpha helix), S (beta sheet), or L (loop).\n",
    "#     Args:\n",
    "#     - top_path (str): The path to the topology file to create the PDB file from.\n",
    "#     - dssp_path (str): The path to save the DSSP file.\n",
    "#     - sim_name (str): Simulation name to be included in the HEADER of the PDB file.\n",
    "#     Returns:\n",
    "#     - List of tuples, each containing the residue number and its secondary structure element.\n",
    "#     \"\"\"\n",
    "#     temp_pdb = \"do_mkdssp.pdb\"\n",
    "\n",
    "#     if sim_name is None:\n",
    "#         sim_name = \"DSSP HEADER\"\n",
    "#     if dssp_path is None:\n",
    "#         dssp_path = \"dssp_file.dssp\"\n",
    "#     print(top_path)\n",
    "#     pdb_test = mda.Universe(top_path)\n",
    "\n",
    "#     # write out as a pdb and add header\n",
    "#     pdb_test.atoms.write(temp_pdb)\n",
    "\n",
    "\n",
    "#     with open(temp_pdb, 'r') as original: data = original.read()\n",
    "#     with open(temp_pdb, 'w') as modified: modified.write('HEADER    '+sim_name+'\\n'+data)\n",
    "\n",
    "#     # Run mkdssp to generate DSSP file\n",
    "#     try:\n",
    "#         subprocess.run(['mkdssp', temp_pdb,  dssp_path], check=True)\n",
    "#     except subprocess.CalledProcessError as e:\n",
    "#         print(f\"Error running DSSP: {e}\")\n",
    "#         return []\n",
    "\n",
    "#     # Parse the DSSP file\n",
    "#     secondary_structures = []\n",
    "#     with open(dssp_path, 'r') as dssp_file:\n",
    "#         # Skip header lines\n",
    "#         for line in dssp_file:\n",
    "#             if line.startswith('  #  RESIDUE AA'):\n",
    "#                 break\n",
    "#         # Read the secondary structure assignments\n",
    "#         for line in dssp_file:\n",
    "#             if len(line) > 13:  # Ensure line has enough data\n",
    "#                 residue_num = line[5:10].strip()\n",
    "#                 ss = line[16]\n",
    "#                 # Simplify the secondary structure to H, S, or L\n",
    "#                 if ss in 'GHI':\n",
    "#                     ss = 'H'  # Helix\n",
    "#                 elif ss in 'EB':\n",
    "#                     ss = 'S'  # Sheet\n",
    "#                 else:\n",
    "#                     ss = 'L'  # Loop or other\n",
    "#                 secondary_structures.append((residue_num, ss))\n",
    "\n",
    "#     # Cleanup temp PDB file\n",
    "#     os.remove(temp_pdb)\n",
    "#     os.remove(dssp_path)\n",
    "#     print(len(secondary_structures))\n",
    "#     print(len(pdb_test.residues))\n",
    "#     return secondary_structures\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HDXER_ENV",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
